Replay mode

franktip opened the pull request:
Adding a replay mode to LLMorpheus to make it possible to replay executions. Documentation has been updated and a test has been added.

max-schaefer commented on README.md:
````diff
@@ -103,8 +103,16 @@ The `StandardStryker.yml` workflow runs the standard StrykerJS mutators.
+node benchmark/createMutants.js --path /Users/xxx/zip-a-folder --mutate "lib/**.ts"  --replay /Users/xxx/recording
+```
+
+A collection of recorded executions of LLMorheus is made available at `https://github.com/neu-se/mutation-testing-data`. Here, you can find recorded executions that were made using the codellama-13b-instruct, codellama-34b-instruct, and mixtral-8x7b-instruct LLMs there, and using various templates and temperature settings. The data is organized in subdirectories that reflect the model, template,  ...
````
Perhaps nicer to format as a link?

```suggestion
A collection of recorded executions of LLMorheus is made available at <https://github.com/neu-se/mutation-testing-data>. Here, you can find recorded executions that were made using the codellama-13b-instruct, codellama-34b-instruct, and mixtral-8x7b-instruct LLMs there, and using various templates and temperature settings. The data is organized in subdirectories that reflect the model, template, and temperature used, e.g. `codellama-13b-instruct/template-full-0.0` reflects a set of experiments using the `codellama-13b-instruct` model, the `full`  template, at temperature `0.0`. This directory contains 5 sets of experiments on 13 subject programs using these settings, contained in subdirectories named `run354`-`run359`. A file `zip/mutants.zip` within each of these subdirectories contains all data that is needed to replay these executions.
```

max-schaefer commented on src/generator/PromptSpecGenerator.ts:
```diff
@@ -7,6 +7,7 @@ import { Prompt } from "../prompt/Prompt";
 import { PromptSpec } from "../prompt/PromptSpec"; 
 import { SourceLocation } from "../util/SourceLocation"; 
 import { charAtPosition, nextPosition, prevPosition } from "../util/code-utils"; 
+import { Completion } from "../prompt/Completion";
```
Why do we need this import?

max-schaefer commented on src/model/ReplayModel.ts:
```diff
@@ -0,0 +1,78 @@
+export class ReplayModel implements IModel {
+
+  private metaInfo: MetaInfo;
+  private promptToCompletionMap: Map<string, string> = new Map<string, string>();
```
I don't think you need the type annotations since the `new` is fully annotated:

```suggestion
  private promptToCompletionMap = new Map<string, string>();
```

max-schaefer reviewed (approved):
I don't know the code very well, but at a high level this looks fine.

I do wonder about all the stuff in `test/input`, though. For example, do we really want to check in a copy of the source code of `zip-a-folder`? Wouldn't it be better to check it out from its own repo on demand?

franktip commented:
I agree with your comments about not including zip-a-folder's source and that it's better to avoid the potential flakiness associated with downloading it as the tests run. I will work on replacing zip-a-folder with a small test project. Thanks!

franktip commented:
I rewrote the test to use a simple example project and removed the dependence on zip-a-folder. I also rewrote several other tests to avoid a dependency on other projects. Please let me know if you have any further comments.

jon-bell commented:
LGTM
